[material-ui][Alert] Deprecate components and componentsProps props

DiegoAndai opened the pull request:
Deprecate `components` and `componentsProps` props in favor of `slots` and `slotProps` respectively.

I wondered whether to deprecate `action`, `icon`, and `iconMappings` as well, but I'm not so sure, and it would be a bigger change, so I created a separate issue to discuss it eventually: https://github.com/mui/material-ui/issues/40680

siriwatknp commented on packages/mui-material/src/Alert/Alert.d.ts:
```diff
@@ -4,13 +4,18 @@ import { SxProps } from '@mui/system';
 
 export interface AlertPropsColorOverrides {}
 
+export interface AlertCloseButtonSlotPropsOverrides {}
+
+export interface AlertCloseIconSlotPropsOverrides {}
```
Do we need to create these interfaces?

DiegoAndai replied:
It's used if users want to extend the respective slot's props with custom props. Do you think we shouldn't create them?

siriwatknp replied:
We can discuss it separately but I would not add it in deprecation PRs (At least, I haven't come across a use case where I have to extend these types before). cc @mnajdova

mnajdova replied:
Agree, let's leave it for later. For now, let's just deprecate and use the same types as before.

DiegoAndai replied:
Done 😃

siriwatknp commented on packages/mui-material/src/Alert/Alert.js:
```diff
@@ -167,11 +168,33 @@ const Alert = React.forwardRef(function Alert(inProps, ref) {
-  const closeIconProps = slotProps.closeIcon ?? componentsProps.closeIcon;
+  const [CloseButtonSlot, closeButtonProps] = useSlot('closeButton', {
+    elementType: IconButton,
+    externalForwardedProps: {
+      slots: backwardCompatibleSlots,
+      slotProps: backwardCompatibleSlotProps,
+    },
```
What do you think about moving `externalForwardedProps` above?

```js
const externalForwardedProps = {
    slots: {
      closeButton: components.CloseButton,
      closeIcon: components.CloseIcon,
      ...slots,
   },
   slotProps: {
     ...componentsProps,
     ...slotProps,
   },
}

```

DiegoAndai replied:
Much better https://github.com/mui/material-ui/pull/40681/commits/2a7780b8b36eba73ffbb3c2114ca9904ee5b99c6 🙌🏼

DiegoAndai commented:
Hey @siriwatknp! I updated this PR. I added the codemods and migration guide. For the codemods, I created a `replaceComponentsWithSlots` deprecations util, as the codemod should be the same for all `components` and `componentsProps` deprecations.

siriwatknp commented on packages/mui-material/src/Alert/Alert.test.js:
```diff
@@ -20,13 +20,14 @@ describe('<Alert />', () => {
     },
-    skip: [
-      'componentsProp',
-      'slotPropsCallback', // not supported yet
```
👍

siriwatknp reviewed (approved):
👏 Awesome
